### ISSUE TITLE
Respect time zones in ICS files

### ISSUE THREAD

**Issue by chrysn** [Wilm0r/giggity#182]
When a conference time zone is not the local one (as has become typical in the last year for reasons), event times and even the "current time" indication of the block schedule become inaccurate.

For an example, load <https://datatracker.ietf.org/meeting/111/agenda.ics>. Right now, the block schedule indicates that "dispatch" starts at 12:00 (which is kind of correct -- Los Angeles time), but it also displays the current time as well into dispatch (13:30), which is inconsistent with the calendar showing LA time.

As far as I can read iCalendar files, the data appears to be correct; the LA time zone is defined at the top with appropriate tzoffsets, and all DTSTART/DTEND are given with with the appropriate TZID and then in the appropriate time. Example:

```
DTSTART;TZID=America/Los_Angeles:20210726T120000
DTEND;TZID=America/Los_Angeles:20210726T140000
```

For a simple enhancement, it'd already help if the current time indication were fixed (so that when opening the app, you see which slots are on right now); for full time zone support I'm afraid this might need some GUI to select whether a conference is shown in its time zone (useful for planning ahead while you're at home before you go there) or in the device's local time zone (useful for attending remotely).

**Comment by Wilm0r** [comment]
I'll have to make some time to study this more accurately.. I did spend some time last year to make Giggity more useful with the remote conferences that come with our current "exciting" time, see 12f60c7f26c23289b966f3c6214add6f0a0df24f. It mentions Pentabarf only because (IIRC) Pentabarf files have a single timezone in the header of the file.

Possibly what's missing is grabbing *the* timezone for an ICS file as long as there is just one. And I *think* the limited ICS parser of Giggity does parse timezone *offsets* if they're there, but not yet timezone strings (which is easier to do when there's a single timezone setting for the whole file, instead of having to re-initialise a timezone-aware parser for every time field you read).

Can you confirm that, other than there not being a GUI, the timezone awareness for Pentabarf files is sensible/usable to you?

**Comment by chrysn** [comment]
I don't have pentabarf-style files for IETF.

Looking at MiniDebConf India (which is pentabarf style), the times are shown correctly, indicating a start of sessions at 06:30 in my local time zone (whereas the sessions started 11:00 Indian time which is +05:30). I can't verify that the green current-time indicator and notifications work (for lack of a current pentabarf conference), but things'd have to be quite odd for this not to work. Things adjust instantly when I switch around time zones on my phone settings (that is, after re-entering the event view; switching back after a TZ change moves me to the giggity main menu which is probably a good thing).

(In contrast, with the ICS calendar of IETF, changing local device time zones has no effect on the displayed start and end times, but the "current time" marker gets moved around, and is accurate relative to the slots when I select the conference time zone as my phone's time zone).

**Comment by chrysn** [comment]
Brief correction from a recent observation from the ICS-based IETF calendar: The notification about imminent meetings *does* arrive at the right time: I see a "in 11 min" notification in the tray about an event that, in the block schedule, has long been switched to the salmon orange of past events.

**Comment by Wilm0r** [comment]
> Things adjust instantly when I switch around time zones on my phone settings (that is, after re-entering the event view; switching back after a TZ change moves me to the giggity main menu which is probably a good thing).

Heh, yeah that's my stop gap fix ac3aabcfa1ad23fefaef28b738a6aa6f26c09a49 at work. Once timezone awareness is complete I might in fact be able to deal with tz changes a little more subtly again. :)

Thanks for the details on how the ICS is going wrong. Can't really get it fixed before the IETF week ends, but hopefully for future events.

**Comment by Wilm0r** [comment]
a079f481 fixed timezone conversions I Believ.
